verilog/tools/kythe/scope_resolver.h

```diff
@@ -0,0 +1,152 @@
+};
+
+// Container with a stack of Scopes to hold the accessible scopes during
+// traversing an Indexing Facts Tree.
+// This is used to get the definitions of some reference.
+//
+// This is modified during tree traversal because in case of entering new
+// scope the new scope is resolved first and after that it's added to the
+// containing scope and the next scope is being analyzed.
+class VerticalScopeResolver : public verible::AutoPopStack<Scope*> {
+ public:
+  typedef verible::AutoPopStack<Scope*> base_type;
+
+  // member class to handle push and pop of stack safely
+  using AutoPop = base_type::AutoPop;
+
+  // returns the top VName of the stack
+  Scope& top() { return *ABSL_DIE_IF_NULL(base_type::top()); }
+
+  // TODO(minatoma): improve performance and memory for this function.
+  //
+  // This function uses string matching to find the definition of some
+  // variable in reverse order of the current scopes.
+  //
+  // Improvement can be replacing the string matching to comparison based on
+  // integers or enums and reshaping the scope to be one vector instead of
+  // vector of vectors.
+  //
+  // TODO(minatoma): consider using vector<pair<name, type>> for signature.
+  //
+  // Search function to get the VName of a definitions of some reference.
+  // It loops over the scopes in reverse order and loops over every scope in
+  // reverse order to find a definition for the variable with given prefix
+  // signature.
+  // e.g
+  // {
+  //    bar#module,
+  //    foo#module,
+  // }
+  // {
+  //    other scope,
+  // }
+  // Given bar#module it return the whole VName of that definition.
+  // And if more than one match is found the first would be returned.
+  const VName* SearchForDefinition(absl::string_view name) const;
+};
+
+// Keeps track and saves the explored scopes with a <key, value> and maps every
+// signature to its scope.
+class HorizontalScopeResolver {
+ public:
+  // Searches for a VName with the given name in the scope with the given
+  // signature.
+  const VName* SearchForVNameInScope(const Signature& signature,
+                                     absl::string_view name) const;
+
+  // Searches for a scope with the given signature in the scopes.
+  const Scope* SearchForScope(const Signature& signature) const;
+
+  // Maps the given signature to the given scope.
+  void MapSignatureToScope(const Signature& signature, const Scope& scope);
+
+  // Maps the given signature to the scope of the other signature.
+  void MapSignatureToScopeOfSignature(const Signature& signature,
+                                      const Signature& other_signature);
+
+ private:
+  // Saved packages signatures alongside with their inner members.
+  // This is used for resolving references to some variables after using
+  // import pkg::*. e.g package pkg1;
+  //   function my_fun(); endfunction
+  //   class my_class; endclass
+  // endpackage
+  //
+  // package pkg2;
+  //   function my_fun(); endfunction
```

Comment:
should this example have an `import pkg1::*;`?

Reply:
Can you add an example that uses `import`?  Or something that illustrates the impact of an import statement?